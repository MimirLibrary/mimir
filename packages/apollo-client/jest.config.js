module.exports = {
  displayName: 'apollo-client',
  preset: '../../jest.preset.js',
  coverageDirectory: '../../coverage/packages/apollo-client',
};
