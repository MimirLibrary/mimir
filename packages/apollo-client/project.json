{
  "sourceRoot": "packages/apollo-client/src",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/apollo-client",
        "main": "packages/apollo-client/src/index.ts",
        "tsConfig": "packages/apollo-client/tsconfig.lib.json",
        "assets": ["packages/apollo-client/*.md"]
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/apollo-client/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/apollo-client"],
      "options": {
        "jestConfig": "packages/apollo-client/jest.config.js",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
