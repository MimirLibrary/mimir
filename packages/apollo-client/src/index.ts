export * from './lib/instance';
export * from './lib/generated/client-types';
