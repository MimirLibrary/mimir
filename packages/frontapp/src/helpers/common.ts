export const shortenText = (text: string, numberToShorten: number) => {
  return text.trim().slice(0, numberToShorten);
};
