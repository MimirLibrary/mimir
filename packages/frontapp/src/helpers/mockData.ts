export const mockStatus = [
  {
    person_id: 1,
    status: 'Free',
    created_at: '2022-09-08T03:58:39.228Z',
    id: 1,
    material_id: 1,
  },
];

export const mockMaterial = [
  {
    __typename: 'Material',
    author: 'test Author',
    category: 'test Category',
    created_at: '2022-09-08T03:58:39.228Z',
    description: 'Test Description',
    id: '1',
    id_type: 'Test Id_Type',
    identifier: '1111111111',
    is_donated: false,
    notification: [],
    picture: '',
    statuses: mockStatus,
    title: 'test',
    type: 'Book',
    updated_at: '2022-09-08T03:58:39.228Z',
  },
];
