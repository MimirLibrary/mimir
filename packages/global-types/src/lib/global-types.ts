
/*
 * -------------------------------------------------------
 * THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
 * -------------------------------------------------------
 */

/* tslint:disable */
/* eslint-disable */
export enum Permissions {
    GRANT_REVOKE_MANAGER = "GRANT_REVOKE_MANAGER"
}

export interface CreateStateInput {
    person_id: number;
    state: boolean;
    description?: Nullable<string>;
}

export interface BookInput {
    identifier: string;
    person_id: number;
}

export interface ProlongTimeInput {
    person_id: number;
    material_id: number;
}

export interface CreateLocationInput {
    location: string;
}

export interface RemoveLocationInput {
    location_id: number;
}

export interface CreateMaterialInput {
    identifier: string;
    id_type: string;
    location_id: number;
    type: string;
    title: string;
    picture?: Nullable<string>;
    author: string;
    category: string;
    description: string;
}

export interface RemoveMaterialInput {
    identifier: string;
    location_id: number;
    type: string;
}

export interface UpdateMaterialInput {
    identifier?: Nullable<string>;
    id_type?: Nullable<string>;
    location_id?: Nullable<number>;
    type?: Nullable<string>;
    title?: Nullable<string>;
    picture?: Nullable<string>;
    author?: Nullable<string>;
    category?: Nullable<string>;
    updated_at: DateTime;
    description?: Nullable<string>;
}

export interface SearchOneMaterial {
    identifier: string;
    locations?: Nullable<number[]>;
}

export interface DonateBookInput {
    identifier: string;
    id_type: string;
    location_id: number;
    type: string;
    title: string;
    author: string;
    category: string;
    description?: Nullable<string>;
    picture?: Nullable<string>;
    person_id: number;
    role: string;
    is_donated: boolean;
}

export interface SearchInput {
    search?: Nullable<string>;
    locations?: Nullable<number[]>;
}

export interface CreateMessageInput {
    title: string;
    message: string;
    material_id?: Nullable<number>;
    person_id: number;
    location_id: number;
}

export interface CreateAnswerNotification {
    id: number;
    person_id: number;
    message: string;
}

export interface CreateNotificationInput {
    material_id: number;
    person_id: number;
}

export interface CreateSimpleNotification {
    person_id: number;
    message: string;
}

export interface RemoveNotificationInput {
    material_id: number;
    person_id: number;
}

export interface CreatePersonInput {
    smg_id: string;
    location_id: number;
    type: string;
}

export interface UpdatePersonLocationInput {
    person_id: number;
    location_id: number;
}

export interface CreateStatusInput {
    material_id: number;
    person_id: number;
    status: string;
}

export interface BlockedUsers {
    id: string;
    description?: Nullable<string>;
    state: boolean;
    created_at: DateTime;
    person: Person;
}

export interface IQuery {
    getBlocksByPerson(person_id: string): Nullable<Nullable<BlockedUsers>[]> | Promise<Nullable<Nullable<BlockedUsers>[]>>;
    getReasonOfBlock(person_id: string): Nullable<BlockedUsers> | Promise<Nullable<BlockedUsers>>;
    getAllTakenItems(person_id: number): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    getItemsForClaimHistory(person_id: number): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    getAllLocations(): Nullable<Location>[] | Promise<Nullable<Location>[]>;
    getAllMaterials(locations?: Nullable<number[]>, limit?: Nullable<number>, offset?: Nullable<number>): Nullable<Material>[] | Promise<Nullable<Material>[]>;
    getMaterialById(id: string): Material | Promise<Material>;
    searchOfMaterials(input: SearchInput): Nullable<Nullable<Material>[]> | Promise<Nullable<Nullable<Material>[]>>;
    getMaterialByIdentifier(input: SearchOneMaterial): Material | Promise<Material>;
    getMaterialByIdentifierFromMetadata(identifier: string): Nullable<IMetaOfMaterial> | Promise<Nullable<IMetaOfMaterial>>;
    getAllDonatedMaterialsByPerson(id: string): Nullable<Nullable<Material>[]> | Promise<Nullable<Nullable<Material>[]>>;
    getMessagesByPerson(person_id: string): Nullable<Nullable<Message>[]> | Promise<Nullable<Nullable<Message>[]>>;
    getAllMessages(location_id: number): Nullable<Message[]> | Promise<Nullable<Message[]>>;
    getNotificationsByPerson(person_id: number): Nullable<Notification>[] | Promise<Nullable<Notification>[]>;
    getNotificationsByMaterial(material_id: number): Nullable<Notification>[] | Promise<Nullable<Notification>[]>;
    getOnePerson(id: string): Person | Promise<Person>;
    getAllPersons(username?: Nullable<string>, locations?: Nullable<number[]>): Person[] | Promise<Person[]>;
    getStatusesByPerson(person_id: string): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    getStatusesByMaterial(material_id: string): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    getAllStatusesIsOverdue(locations?: Nullable<number[]>): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    welcome(): string | Promise<string>;
}

export interface IMutation {
    createState(input: CreateStateInput): Nullable<BlockedUsers> | Promise<Nullable<BlockedUsers>>;
    claimBook(input?: Nullable<BookInput>): BookUnionResult | Promise<BookUnionResult>;
    returnItem(input?: Nullable<BookInput>): BookUnionResult | Promise<BookUnionResult>;
    rejectItem(input?: Nullable<BookInput>): BookUnionResult | Promise<BookUnionResult>;
    prolongClaimPeriod(input?: Nullable<ProlongTimeInput>): BookUnionResult | Promise<BookUnionResult>;
    createLocation(input: CreateLocationInput): Nullable<Location> | Promise<Nullable<Location>>;
    removeLocation(input: RemoveLocationInput): Nullable<Location> | Promise<Nullable<Location>>;
    createMaterial(input: CreateMaterialInput): Material | Promise<Material>;
    removeMaterial(input: RemoveMaterialInput): Material | Promise<Material>;
    updateMaterial(input: UpdateMaterialInput): Material | Promise<Material>;
    donateBook(input: DonateBookInput): Material | Promise<Material>;
    createMessageForManager(input: CreateMessageInput): Message | Promise<Message>;
    createNotification(input: CreateNotificationInput): Nullable<Notification> | Promise<Nullable<Notification>>;
    removeNotification(input: RemoveNotificationInput): Nullable<Notification> | Promise<Nullable<Notification>>;
    createAnswerNotification(input: CreateAnswerNotification): Notification | Promise<Notification>;
    createSimpleNotification(input?: Nullable<CreateSimpleNotification>): Notification | Promise<Notification>;
    createPerson(input: CreatePersonInput): Person | Promise<Person>;
    addPersonLocation(input: UpdatePersonLocationInput): Person | Promise<Person>;
    removePersonLocation(input: UpdatePersonLocationInput): Person | Promise<Person>;
    changePersonRole(person_id: number, type: string): Person | Promise<Person>;
    createStatus(input: CreateStatusInput): Status | Promise<Status>;
}

export interface Location {
    id: string;
    location: string;
    persons?: Nullable<Nullable<Person>[]>;
    materials?: Nullable<Nullable<Material>[]>;
}

export interface Material {
    id: string;
    identifier: string;
    id_type: string;
    type: string;
    location_id: number;
    created_at: DateTime;
    updated_at: DateTime;
    title: string;
    picture?: Nullable<string>;
    author: string;
    category: string;
    location: Location;
    statuses: Nullable<Status>[];
    notifications: Nullable<Notification>[];
    messages: Nullable<Message>[];
    description: string;
    is_donated: boolean;
}

export interface ResponseMetadata {
    idType: string;
    value: string;
}

export interface Author {
    id?: Nullable<string>;
    name?: Nullable<string>;
}

export interface Publisher {
    id?: Nullable<string>;
    name?: Nullable<string>;
}

export interface Meta {
    ageRestriction?: Nullable<string>;
    coverType?: Nullable<string>;
    dimensions?: Nullable<string>;
    manufacturer?: Nullable<string>;
    mass?: Nullable<string>;
    numberOfPages?: Nullable<string>;
    price?: Nullable<string>;
    series?: Nullable<string>;
    sku?: Nullable<string>;
}

export interface IMaterialMeta {
    authors?: Nullable<Nullable<Author>[]>;
    cover?: Nullable<string>;
    description?: Nullable<string>;
    title?: Nullable<string>;
    yearPublishedAt?: Nullable<number>;
    meta?: Nullable<Meta>;
    publisher?: Nullable<Publisher>;
}

export interface IMetaOfMaterial {
    idType?: Nullable<string>;
    value?: Nullable<string>;
    material?: Nullable<IMaterialMeta>;
}

export interface Message {
    id: string;
    material_id?: Nullable<number>;
    person_id: number;
    created_at: DateTime;
    person: Person;
    material: Material;
    title: string;
    message: string;
    location_id: number;
}

export interface Notification {
    id: string;
    material_id?: Nullable<number>;
    person_id: number;
    created_at: DateTime;
    person: Person;
    material?: Nullable<Material>;
    message?: Nullable<string>;
}

export interface Person {
    id: string;
    smg_id: string;
    position: string;
    type: string;
    username: string;
    email: string;
    avatar: string;
    created_at: DateTime;
    permissions?: Nullable<Nullable<Permissions>[]>;
    statuses?: Nullable<Nullable<Status>[]>;
    notifications?: Nullable<Nullable<Notification>[]>;
    location?: Nullable<Location[]>;
    messages?: Nullable<Nullable<Message>[]>;
    states?: Nullable<Nullable<BlockedUsers>[]>;
}

export interface Status {
    id: string;
    material_id: number;
    person_id: number;
    status: string;
    created_at: DateTime;
    material: Material;
    person: Person;
}

export interface Error {
    message: string;
}

export type DateTime = any;
export type Time = any;
export type Timestamp = any;
export type TimeZone = any;
export type UtcOffset = any;
export type Duration = any;
export type ISO8601Duration = any;
export type LocalDate = any;
export type LocalTime = any;
export type LocalEndTime = any;
export type EmailAddress = any;
export type NegativeFloat = any;
export type NegativeInt = any;
export type NonEmptyString = any;
export type NonNegativeFloat = any;
export type NonNegativeInt = any;
export type NonPositiveFloat = any;
export type NonPositiveInt = any;
export type PhoneNumber = any;
export type PositiveFloat = any;
export type PositiveInt = any;
export type PostalCode = any;
export type UnsignedFloat = any;
export type UnsignedInt = any;
export type URL = any;
export type BigInt = any;
export type Long = any;
export type Byte = any;
export type UUID = any;
export type GUID = any;
export type Hexadecimal = any;
export type HexColorCode = any;
export type HSL = any;
export type HSLA = any;
export type IPv4 = any;
export type IPv6 = any;
export type ISBN = any;
export type JWT = any;
export type Latitude = any;
export type Longitude = any;
export type MAC = any;
export type Port = any;
export type RGB = any;
export type RGBA = any;
export type SafeInt = any;
export type USCurrency = any;
export type Currency = any;
export type JSON = any;
export type JSONObject = any;
export type IBAN = any;
export type ObjectID = any;
export type Void = any;
export type DID = any;
export type CountryCode = any;
export type Locale = any;
export type RoutingNumber = any;
export type AccountNumber = any;
export type BookUnionResult = Status | Error;
type Nullable<T> = T | null;
