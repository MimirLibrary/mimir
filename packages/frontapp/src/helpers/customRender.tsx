import '@testing-library/jest-dom';
import { render, RenderOptions } from '@testing-library/react';
import React, { FC, ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { ApolloProvider } from '@apollo/client';
import { client } from '@mimir/apollo-client';

jest.mock('react-i18next', () => ({
  // this mock makes sure any components using the translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str: string | null) => str,
      i18n: {
        changeLanguage: () =>
          new Promise(() => {
            return;
          }),
      },
    };
  },
}));

jest.unmock('react-redux');

const wrapper: FC<{ children: React.ReactNode }> = ({ children }) => {
  return (
    <BrowserRouter>
      <ApolloProvider client={client}>
        <Provider store={store}>{children}</Provider>
      </ApolloProvider>
    </BrowserRouter>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => {
  return render(ui, { wrapper, ...options });
};

export { customRender as render };
