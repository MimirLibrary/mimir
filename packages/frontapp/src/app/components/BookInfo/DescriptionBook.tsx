import React, { FC, useEffect, useMemo } from 'react';
import bookImage from '../../../assets/MOC-data/BookImage.png';
import { WrapperInput } from '../Search';
import { RolesTypes } from '@mimir/global-types';
import { listOfGenres } from '../../../assets/SearchConsts';
import { TUserLocation } from '../../store/slices/userSlice';
import BookStatus from '../BookStatus';
import styled from '@emotion/styled';
import { colors, dimensions } from '@mimir/ui-kit';
import { useAppSelector } from '../../hooks/useTypedSelector';
import Dropdown, { IDropdownOption } from '../Dropdown';
import { useGetAllLocationsQuery } from '@mimir/apollo-client';
import { INewData, Location, OpenLink, TitleHolder, Topic } from './index';
import { toast } from 'react-toastify';

export const WrapperInfo = styled.div`
  display: flex;
  @media (max-width: ${dimensions.phone_width}) {
    flex-direction: column;
    align-items: center;
  }
`;

export const BookImage = styled.img`
  display: inline-block;
  width: 12rem;
  height: 19.5rem;
  border-radius: 10px;
  @media (max-width: ${dimensions.phone_width}) {
    margin-right: ${dimensions.base};
  }
`;

export const ShortDescription = styled.div`
  width: 100%;
  margin-left: ${dimensions.xl_2};
  @media (max-width: ${dimensions.phone_width}) {
    margin: 0;
  }
`;

const StyledInput = styled.input`
  width: 19rem;
  border: none;
  outline: none;
  margin-left: ${dimensions.xs_2};
  color: ${colors.main_black};
  margin-right: 0.12rem;
`;

export const TitleBook = styled.h3`
  font-weight: 700;
  margin-bottom: ${dimensions.xl_2};
  font-size: ${dimensions.xl};
  line-height: ${dimensions.xl_2};
  color: ${colors.main_black};
  @media (max-width: ${dimensions.phone_width}) {
    margin-bottom: 0;
    margin-top: ${dimensions.base};
  }
`;

const WrapperDropDown = styled.div`
  margin-top: ${dimensions.base};
  width: 21.5rem;
`;

const RestyledDropdown = styled(Dropdown)`
  margin-top: ${dimensions.xs_2};
  max-width: 21.5rem;
  width: 100%;
`;

export const TopicDescription = styled.p`
  font-weight: 300;
  font-size: ${dimensions.base};
  line-height: ${dimensions.xl};
  color: ${colors.main_black};
`;

const StyledStatus = styled.div`
  font-size: ${dimensions.base};
`;

const StyledInputDeadline = styled.input`
  border: 0.5px solid #bdbdbd;
  border-radius: ${dimensions.xl_3};
  padding: 10px 0;
  width: 3.7rem;
  outline: none;
  padding-left: ${dimensions.xl};
  background: ${colors.bg_secondary};
  :hover {
    border: 0.5px solid ${colors.accent_color};
  }
  :focus {
    border: 0.5px solid ${colors.accent_color};
  }
  @media (max-width: ${dimensions.tablet_width}) {
    width: 100%;
  }
  @media (max-width: ${dimensions.phone_width}) {
    width: 70%;
  }
  ::-webkit-inner-spin-button,
  ::-webkit-outer-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
`;

interface IDescriptionBook {
  editing: boolean;
  src: string | null;
  title: string;
  category: string | undefined;
  author: string;
  status: string | undefined;
  date: Date;
  location: Location;
  newTitleAndAuthor: INewData;
  newDeadline: number;
  handleChangeDeadline: (e: React.ChangeEvent<HTMLInputElement>) => void;
  handleChangeLocation: (option: TUserLocation) => void;
  handleChangeAuthorAndTitle: (e: React.ChangeEvent<HTMLInputElement>) => void;
  handleChangeNewGenre: (e: IDropdownOption) => void;
}

const DescriptionBook: FC<IDescriptionBook> = ({
  editing,
  src,
  location,
  category,
  date,
  status,
  title,
  author,
  newTitleAndAuthor,
  newDeadline,
  handleChangeLocation,
  handleChangeNewGenre,
  handleChangeDeadline,
  handleChangeAuthorAndTitle,
}) => {
  const { userRole } = useAppSelector((state) => state.user);

  const { data: allLocations, error: errorLocations } = useGetAllLocationsQuery(
    {
      skip: userRole === RolesTypes.READER,
    }
  );

  useEffect(() => {
    if (errorLocations) {
      toast.error(errorLocations.message);
    }
  }, [errorLocations]);

  const currentGenreIndex = useMemo(
    () => listOfGenres.findIndex((genre) => genre.value === category),
    [category, listOfGenres]
  );

  const currentLocationIndex = useMemo(
    () =>
      allLocations?.getAllLocations.findIndex((loc) => loc!.id === location.id),
    [allLocations]
  );

  return (
    <WrapperInfo data-testid="description-book">
      <BookImage src={src || bookImage} />
      <ShortDescription>
        {editing ? (
          <>
            <TitleHolder>Name </TitleHolder>
            <WrapperInput>
              <StyledInput
                type="text"
                value={newTitleAndAuthor.newTitle}
                name="newTitle"
                onChange={handleChangeAuthorAndTitle}
              />
            </WrapperInput>
          </>
        ) : (
          <TitleBook>{title || 'Book Title'}</TitleBook>
        )}
        {!editing && <Topic>Genre: </Topic>}
        {userRole === RolesTypes.READER ? (
          <OpenLink>{category || 'Genres of book'}</OpenLink>
        ) : editing ? (
          <WrapperDropDown>
            <TitleHolder>Genre </TitleHolder>
            <RestyledDropdown
              options={listOfGenres}
              initIndex={currentGenreIndex}
              onChange={handleChangeNewGenre}
              placeholder="Enter genre"
            />
          </WrapperDropDown>
        ) : (
          <TopicDescription>{category || 'Genres of book'}</TopicDescription>
        )}
        {editing ? (
          <>
            <br />
            <TitleHolder>Author </TitleHolder>
            <WrapperInput>
              <StyledInput
                type="text"
                name="newAuthor"
                value={newTitleAndAuthor.newAuthor}
                onChange={handleChangeAuthorAndTitle}
              />
            </WrapperInput>
          </>
        ) : (
          <>
            <Topic>Author: </Topic>
            <TopicDescription>{author || 'Author Name'}</TopicDescription>
          </>
        )}
        <>
          {editing ? (
            <WrapperDropDown>
              <TitleHolder>Location </TitleHolder>
              <RestyledDropdown
                options={allLocations!.getAllLocations.map((loc) => ({
                  id: loc!.id,
                  value: loc!.location,
                }))}
                initIndex={currentLocationIndex}
                onChange={(option) =>
                  handleChangeLocation(option as TUserLocation)
                }
              />
            </WrapperDropDown>
          ) : (
            <>
              <Topic>Location: </Topic>
              <TopicDescription>{location.location}</TopicDescription>
            </>
          )}
        </>
        {userRole === RolesTypes.READER ? (
          <>
            <Topic>State: </Topic>
            <StyledStatus>
              <BookStatus status={status} date={date} />
            </StyledStatus>
          </>
        ) : editing ? (
          <>
            <br />
            <TitleHolder>Deadline </TitleHolder>
            <StyledInputDeadline
              value={newDeadline}
              type="number"
              onChange={handleChangeDeadline}
              min="1"
              max="31"
            />{' '}
            days
          </>
        ) : (
          <>
            <Topic>Deadline: </Topic>
            <TopicDescription>{newDeadline + ' days'}</TopicDescription>
          </>
        )}
      </ShortDescription>
    </WrapperInfo>
  );
};

export default DescriptionBook;
