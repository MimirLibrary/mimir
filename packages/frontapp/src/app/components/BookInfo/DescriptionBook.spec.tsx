import { render } from '../../../helpers/customRender';
import DescriptionBook from './DescriptionBook';
import { screen } from '@testing-library/react';
import * as reactHooks from 'react-redux';
import { RolesTypes } from '@mimir/global-types';
import { MockedProvider } from '@apollo/client/testing';
import { GetAllLocationsDocument } from '@mimir/apollo-client';
import { Location } from './index';

jest.mock('react-redux');

const locationMock = {
  request: {
    query: GetAllLocationsDocument,
  },
  result: {
    data: {
      getAllLocations: [
        { id: '1', location: 'Gomel' },
        { id: '2', location: 'Minsk' },
      ],
    },
  },
};

describe('render DescriptionBook component', () => {
  beforeAll(() => {
    jest.spyOn(reactHooks, 'useSelector').mockReturnValue(RolesTypes.READER);
  });

  it('should be correct description of book', async () => {
    const handleChangeSomeData = jest.fn();
    jest.spyOn(reactHooks, 'useSelector').mockReturnValue(RolesTypes.READER);
    render(
      <MockedProvider mocks={[locationMock]} addTypename={false}>
        <DescriptionBook
          title="Dracula"
          author="Ivan Uglovets"
          editing={false}
          category="Fantasy"
          location={{ id: '2', location: 'Gomel' }}
          date={new Date('2022-09-21T08:29:24.221Z')}
          status="Free"
          src=""
          newTitleAndAuthor={{ newAuthor: 'Ivan', newTitle: 'Spider-Man' }}
          newDeadline={31}
          handleChangeAuthorAndTitle={handleChangeSomeData}
          handleChangeDeadline={handleChangeSomeData}
          handleChangeLocation={handleChangeSomeData}
          handleChangeNewGenre={handleChangeSomeData}
        />
      </MockedProvider>
    );
    expect(screen.getByText('Dracula')).toBeInTheDocument();
    expect(screen.getByText('Ivan Uglovets')).toBeInTheDocument();
    expect(screen.getByText('Gomel')).toBeInTheDocument();
  });

  it('render without some data', async () => {
    const handleChangeSomeData = jest.fn();

    render(
      <MockedProvider mocks={[locationMock]} addTypename={false}>
        <DescriptionBook
          title=""
          author=""
          editing={false}
          category=""
          location={{} as Location}
          date={new Date('2022-09-21T08:29:24.221Z')}
          status="Free"
          src=""
          newTitleAndAuthor={{ newAuthor: 'Ivan', newTitle: 'Spider-Man' }}
          newDeadline={31}
          handleChangeAuthorAndTitle={handleChangeSomeData}
          handleChangeDeadline={handleChangeSomeData}
          handleChangeLocation={handleChangeSomeData}
          handleChangeNewGenre={handleChangeSomeData}
        />
      </MockedProvider>
    );
    expect(screen.queryByText('Dracula')).toBeNull();
    expect(screen.queryByText('Ivan Uglovets')).toBeNull();
    expect(screen.queryByText('Gomel')).toBeNull();
  });

  it('renders correctly with snapshots', () => {
    const handleChangeSomeData = jest.fn();
    jest.spyOn(reactHooks, 'useSelector').mockReturnValue(RolesTypes.READER);
    render(
      <MockedProvider mocks={[locationMock]} addTypename={false}>
        <DescriptionBook
          title="Dracula"
          author="Ivan Uglovets"
          editing={false}
          category="Fantasy"
          location={{ id: '2', location: 'Gomel' }}
          date={new Date('2022-09-21T08:29:24.221Z')}
          status="Free"
          src=""
          newTitleAndAuthor={{ newAuthor: 'Ivan', newTitle: 'Spider-Man' }}
          newDeadline={31}
          handleChangeAuthorAndTitle={handleChangeSomeData}
          handleChangeDeadline={handleChangeSomeData}
          handleChangeLocation={handleChangeSomeData}
          handleChangeNewGenre={handleChangeSomeData}
        />
      </MockedProvider>
    );
    expect(screen.getByTestId('description-book')).toMatchSnapshot();
  });
});
