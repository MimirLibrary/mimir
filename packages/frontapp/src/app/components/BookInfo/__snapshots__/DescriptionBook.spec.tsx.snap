// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render DescriptionBook component renders correctly with snapshots 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media (max-width: 30rem) {
  .emotion-0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

.emotion-1 {
  display: inline-block;
  width: 12rem;
  height: 19.5rem;
  border-radius: 10px;
}

@media (max-width: 30rem) {
  .emotion-1 {
    margin-right: 1rem;
  }
}

.emotion-2 {
  width: 100%;
  margin-left: 1.5rem;
}

@media (max-width: 30rem) {
  .emotion-2 {
    margin: 0;
  }
}

.emotion-3 {
  font-weight: 700;
  margin-bottom: 1.5rem;
  font-size: 1.25rem;
  line-height: 1.5rem;
  color: #333333;
}

@media (max-width: 30rem) {
  .emotion-3 {
    margin-bottom: 0;
    margin-top: 1rem;
  }
}

.emotion-4 {
  margin: 1rem 0 0.5rem 0;
  font-weight: 500;
  font-size: 1rem;
  line-height: 1.25rem;
  color: #333333;
}

.emotion-5 {
  cursor: pointer;
  margin: 0.5rem 0;
  font-weight: 300;
  color: #1A1ED6;
  font-size: 1rem;
  line-height: 1.25rem;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-7 {
  font-weight: 300;
  font-size: 1rem;
  line-height: 1.25rem;
  color: #333333;
}

.emotion-11 {
  font-size: 1rem;
}

@media (max-width: 30rem) {
  .emotion-12 {
    width: -webkit-fit-content;
    width: -moz-fit-content;
    width: fit-content;
    padding: 0 4px;
    border-radius: 4px;
  }
}

.emotion-13 {
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1rem;
  color: #2AA816;
}

@media (max-width: 30rem) {
  .emotion-13 {
    font-size: 0.625rem;
  }
}

<div
  class="emotion-0"
  data-testid="description-book"
>
  <img
    class="emotion-1"
    src="BookImage.png"
  />
  <div
    class="emotion-2"
  >
    <h3
      class="emotion-3"
    >
      Dracula
    </h3>
    <p
      class="emotion-4"
    >
      Genre: 
    </p>
    <a
      class="emotion-5"
    >
      Fantasy
    </a>
    <p
      class="emotion-4"
    >
      Author: 
    </p>
    <p
      class="emotion-7"
    >
      Ivan Uglovets
    </p>
    <p
      class="emotion-4"
    >
      Location: 
    </p>
    <p
      class="emotion-7"
    >
      Gomel
    </p>
    <p
      class="emotion-4"
    >
      State: 
    </p>
    <div
      class="emotion-11"
    >
      <div
        class="emotion-12"
        data-testid="bookStatus"
      >
        <p
          class="emotion-13"
        >
          Statuses.Free
        </p>
      </div>
    </div>
  </div>
</div>
`;
